### PR TITLE
Fix for Issue #1133 with Test

### LINKED ISSUE #1133: Print E**x as exp(x)
This is what sympy does and looks better.

### PR DESCRIPTION
**Pull request by ShikharJ**
Please suggest any necessary changes if required.

Fixes #1133

### CODE REVIEW

**Comment by ShikharJ** [comment]
@isuruf ping.

**Comment by isuruf** [comment]
Can you add some examples like `exp(-x)`, `exp(x) * 10`, `exp(-1)` ?

**Comment by certik** [comment]
@ShikharJ I think that's the right fix. thanks for it. If you could add a few more tests as @isuruf suggested to ensure it works, that would be great.

**Comment by ShikharJ** [comment]
@certik @isuruf I'm on it.

**Comment by isuruf** [comment]
@ShikharJ, with this, I think `exp(-1)` and `exp(-1)*2` are printed differently.
Also with sympy, take a look at the string printer, not the pretty printer.

**Comment by ShikharJ** [comment]
@isuruf Indeed, `exp(-1)*2` is being printed as `2/E`. I'll try to fix that.

**Comment by ShikharJ** [comment]
@isuruf A review?

File: symengine/printer.cpp [hunk 1]
```diff
@@ -391,6 +396,8 @@ void StrPrinter::bvisit(const Mul &x)
         s2 = s2.substr(0, s2.size() - 1);
         if (den > 1) {
             str_ = s + "/(" + s2 + ")";
+        } else if (s2[0] == 'e' and s2[1] == 'x' and s2[2] == 'p') {
```
Comment (certik) [inline symengine/printer.cpp]: Can this run out of bounds if `len(s2) < 3`?

Comment (certik) [inline symengine/printer.cpp]: Cannot this be written simply as `s2.substr(3) == "exp"`?

Comment (ShikharJ) [inline symengine/printer.cpp]: I'll make the changes.

**Comment by ShikharJ** [comment]
@certik Using s2.substr(3) == "exp" is setting off errors in other tests:
```
Test project /home/shikhar/symengine
      Start  1: test_rcp
 1/34 Test  #1: test_rcp .........................   Passed    0.01 sec
      Start  2: test_basic
 2/34 Test  #2: test_basic .......................   Passed    0.04 sec
      Start  3: test_arit
 3/34 Test  #3: test_arit ........................   Passed    0.13 sec
      Start  4: test_poly
 4/34 Test  #4: test_poly ........................   Passed    0.01 sec
      Start  5: test_series
 5/34 Test  #5: test_series ......................   Passed    0.01 sec
      Start  6: test_series_generic
 6/34 Test  #6: test_series_generic ..............   Passed    1.00 sec
      Start  7: test_functions
 7/34 Test  #7: test_functions ...................   Passed    0.10 sec
      Start  8: test_subs
 8/34 Test  #8: test_subs ........................   Passed    0.01 sec
      Start  9: test_integer
 9/34 Test  #9: test_integer .....................   Passed    0.01 sec
      Start 10: test_rational
10/34 Test #10: test_rational ....................   Passed    0.01 sec
      Start 11: test_number
11/34 Test #11: test_number ......................   Passed    0.01 sec
      Start 12: test_as_numer_denom
12/34 Test #12: test_as_numer_denom ..............   Passed    0.01 sec
      Start 13: test_parser
13/34 Test #13: test_parser ......................   Passed    0.02 sec
      Start 14: test_sets
14/34 Test #14: test_sets ........................   Passed    0.01 sec
      Start 15: test_fields
15/34 Test #15: test_fields ......................   Passed    0.14 sec
      Start 16: test_infinity
16/34 Test #16: test_infinity ....................   Passed    0.01 sec
      Start 17: test_ntheory
17/34 Test #17: test_ntheory .....................   Passed    0.08 sec
      Start 18: test_diophantine
18/34 Test #18: test_diophantine .................   Passed    0.03 sec
      Start 19: test_matrix
19/34 Test #19: test_matrix ......................   Passed    0.17 sec
      Start 20: test_printing
20/34 Test #20: test_printing ....................***Failed    0.01 sec
      Start 21: test_ccode
21/34 Test #21: test_ccode .......................   Passed    0.01 sec
      Start 22: test_uintpoly
22/34 Test #22: test_uintpoly ....................   Passed    0.01 sec
      Start 23: test_uratpoly
23/34 Test #23: test_uratpoly ....................   Passed    0.01 sec
      Start 24: test_mintpoly
24/34 Test #24: test_mintpoly ....................   Passed    0.01 sec
      Start 25: test_uexprpoly
25/34 Test #25: test_uexprpoly ...................   Passed    0.02 sec
      Start 26: test_mexprpoly
26/34 Test #26: test_mexprpoly ...................***Failed    0.04 sec
      Start 27: test_basic_conversions
27/34 Test #27: test_basic_conversions ...........   Passed    0.03 sec
      Start 28: test_evalf
28/34 Test #28: test_evalf .......................   Passed    0.01 sec
      Start 29: test_eval_double
29/34 Test #29: test_eval_double .................   Passed    0.01 sec
      Start 30: test_lambda_double
30/34 Test #30: test_lambda_double ...............   Passed    0.01 sec
      Start 31: test_cwrapper
31/34 Test #31: test_cwrapper ....................***Exception: Other  0.15 sec
      Start 32: test_expression
32/34 Test #32: test_expression ..................   Passed    0.03 sec
      Start 33: test_logic
33/34 Test #33: test_logic .......................   Passed    0.01 sec
      Start 34: test_finitediff
34/34 Test #34: test_finitediff ..................   Passed    0.02 sec

91% tests passed, 3 tests failed out of 34

Total Test time (real) =   2.21 sec

The following tests FAILED:
	 20 - test_printing (Failed)
	 26 - test_mexprpoly (Failed)
	 31 - test_cwrapper (OTHER_FAULT)
Errors while running CTest
```

**Comment by certik** [comment]
Post the output of `ctest --output-on-failure`, also make sure you are in Debug mode.

**Comment by ShikharJ** [comment]
@certik 
```
Test project /home/shikhar/symengine
      Start  1: test_rcp
 1/34 Test  #1: test_rcp .........................   Passed    0.01 sec
      Start  2: test_basic
 2/34 Test  #2: test_basic .......................   Passed    0.04 sec
      Start  3: test_arit
 3/34 Test  #3: test_arit ........................   Passed    0.13 sec
      Start  4: test_poly
 4/34 Test  #4: test_poly ........................   Passed    0.02 sec
      Start  5: test_series
 5/34 Test  #5: test_series ......................   Passed    0.02 sec
      Start  6: test_series_generic
 6/34 Test  #6: test_series_generic ..............   Passed    1.00 sec
      Start  7: test_functions
 7/34 Test  #7: test_functions ...................   Passed    0.10 sec
      Start  8: test_subs
 8/34 Test  #8: test_subs ........................   Passed    0.01 sec
      Start  9: test_integer
 9/34 Test  #9: test_integer .....................   Passed    0.01 sec
      Start 10: test_rational
10/34 Test #10: test_rational ....................   Passed    0.01 sec
      Start 11: test_number
11/34 Test #11: test_number ......................   Passed    0.01 sec
      Start 12: test_as_numer_denom
12/34 Test #12: test_as_numer_denom ..............   Passed    0.01 sec
      Start 13: test_parser
13/34 Test #13: test_parser ......................   Passed    0.02 sec
      Start 14: test_sets
14/34 Test #14: test_sets ........................   Passed    0.01 sec
      Start 15: test_fields
15/34 Test #15: test_fields ......................   Passed    0.14 sec
      Start 16: test_infinity
16/34 Test #16: test_infinity ....................   Passed    0.01 sec
      Start 17: test_ntheory
17/34 Test #17: test_ntheory .....................   Passed    0.08 sec
      Start 18: test_diophantine
18/34 Test #18: test_diophantine .................   Passed    0.02 sec
      Start 19: test_matrix
19/34 Test #19: test_matrix ......................   Passed    0.13 sec
      Start 20: test_printing
20/34 Test #20: test_printing ....................***Failed    0.02 sec

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_printing is a Catch v1.1 b14 (develop) host application.
Run with -? for options

-------------------------------------------------------------------------------
test_printing(): printing
-------------------------------------------------------------------------------
/home/shikhar/symengine/symengine/tests/printing/test_printing.cpp:58
...............................................................................

/home/shikhar/symengine/symengine/tests/printing/test_printing.cpp:93: FAILED:
  REQUIRE( r->__str__() == "2*exp(-1)" )
with expansion:
  "2/exp(-1)" == "2*exp(-1)"

 _____           _____         _         
|   __|_ _ _____|   __|___ ___|_|___ ___ 
|__   | | |     |   __|   | . | |   | -_|
|_____|_  |_|_|_|_____|_|_|_  |_|_|_|___|
      |___|               |___|          

===============================================================================
test cases:  9 |  8 passed | 1 failed
assertions: 55 | 54 passed | 1 failed


      Start 21: test_ccode
21/34 Test #21: test_ccode .......................   Passed    0.01 sec
      Start 22: test_uintpoly
22/34 Test #22: test_uintpoly ....................   Passed    0.01 sec
      Start 23: test_uratpoly
23/34 Test #23: test_uratpoly ....................   Passed    0.01 sec
      Start 24: test_mintpoly
24/34 Test #24: test_mintpoly ....................   Passed    0.02 sec
      Start 25: test_uexprpoly
25/34 Test #25: test_uexprpoly ...................   Passed    0.02 sec
      Start 26: test_mexprpoly
26/34 Test #26: test_mexprpoly ...................***Failed    0.04 sec

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_mexprpoly is a Catch v1.1 b14 (develop) host application.
Run with -? for options

-------------------------------------------------------------------------------
Testing MExprPoly::eval
-------------------------------------------------------------------------------
/home/shikhar/symengine/symengine/tests/polynomial/test_mexprpoly.cpp:142
...............................................................................

/home/shikhar/symengine/symengine/tests/polynomial/test_mexprpoly.cpp:177: FAILED:
  REQUIRE( p->eval(m1) == expr4 )
due to unexpected exception with message:
  basic_string::substr: __pos (which is 3) > this->size() (which is 1)

===============================================================================
test cases: 16 | 15 passed | 1 failed
assertions: 88 | 87 passed | 1 failed


      Start 27: test_basic_conversions
27/34 Test #27: test_basic_conversions ...........   Passed    0.03 sec
      Start 28: test_evalf
28/34 Test #28: test_evalf .......................   Passed    0.01 sec
      Start 29: test_eval_double
29/34 Test #29: test_eval_double .................   Passed    0.01 sec
      Start 30: test_lambda_double
30/34 Test #30: test_lambda_double ...............   Passed    0.01 sec
      Start 31: test_cwrapper
31/34 Test #31: test_cwrapper ....................***Exception: Other  0.15 sec
terminate called after throwing an instance of 'std::out_of_range'
  what():  basic_string::substr: __pos (which is 3) > this->size() (which is 1)

      Start 32: test_expression
32/34 Test #32: test_expression ..................   Passed    0.02 sec
      Start 33: test_logic
33/34 Test #33: test_logic .......................   Passed    0.01 sec
      Start 34: test_finitediff
34/34 Test #34: test_finitediff ..................   Passed    0.02 sec

91% tests passed, 3 tests failed out of 34

Total Test time (real) =   2.19 sec

The following tests FAILED:
	 20 - test_printing (Failed)
	 26 - test_mexprpoly (Failed)
	 31 - test_cwrapper (OTHER_FAULT)
Errors while running CTest
```

**Comment by certik** [comment]
I see, the line should have been:
```c++
s2.substr(0, 3) == "exp"
```
See http://en.cppreference.com/w/cpp/string/basic_string/substr

File: symengine/printer.cpp [hunk 2]
```diff
@@ -357,7 +359,7 @@ void StrPrinter::bvisit(const Mul &x)
     }
 
     for (const auto &p : x.dict_) {
-        if ((is_a<Integer>(*p.second)
+        if ((is_a<Integer>(*p.second) and neq(*(p.first), *E)
```
Comment (isuruf) [inline symengine/printer.cpp]: This should be for the whole `||` not just integer condition.

File: symengine/printer.cpp [hunk 3]
```diff
@@ -334,8 +334,10 @@ void StrPrinter::bvisit(const Add &x)
 void StrPrinter::_print_pow(std::ostringstream &o, const RCP<const Basic> &a,
                             const RCP<const Basic> &b)
 {
-    if (eq(*b, *rational(1, 2))) {
+    if (eq(*b, *rational(1, 2)) and neq(*a, *E)) {
         o << "sqrt(" << apply(a) << ")";
+    } else if (eq(*a, *E)) {
```
Comment (isuruf) [inline symengine/printer.cpp]: Can you make this the first condition, so that it is simpler ?

File: symengine/printer.cpp (hunk 2 above)
Comment (ShikharJ) [inline symengine/printer.cpp]: @isuruf Like this?
```
if (((is_a<Integer>(*p.second)
             and rcp_static_cast<const Integer>(p.second)->is_negative())
            || (is_a<Rational>(*p.second)
                and rcp_static_cast<const Rational>(p.second)->is_negative())) and neq(*(p.first), *E))
```

Comment (isuruf) [inline symengine/printer.cpp]: Yes, you can simplify this a bit by doing,
```C++
if ((is_a<Integer>(*p.second) or is_a<Rational>(*p.second)) and static_cast<const Number &>(*p.second).is_negative()) and neq(*p.first, *E))
```

**Pull request merged**